sklearnex/neighbors/_lof.py

```diff
@@ -0,0 +1,167 @@
+from .common import KNeighborsDispatchingBase
+from .knn_unsupervised import NearestNeighbors
+
+
+@control_n_jobs(decorated_methods=["fit", "kneighbors"])
+class LocalOutlierFactor(KNeighborsDispatchingBase, sklearn_LocalOutlierFactor):
+    __doc__ = (
+        sklearn_LocalOutlierFactor.__doc__
+        + "\n NOTE: When X=None, methods kneighbors, kneighbors_graph, and predict will"
+        + "\n only output numpy arrays. In that case, the only way to offload to gpu"
+        + "\n is to use a global queue (e.g. using config_context)"
+    )
+    if sklearn_check_version("1.2"):
+        _parameter_constraints: dict = {
+            **sklearn_LocalOutlierFactor._parameter_constraints
+        }
+
+    # Only certain methods should be taken from knn to prevent code
+    # duplication. Inheriting would yield a complicated inheritance
+    # structure and violate the sklearn inheritance path.
+    _save_attributes = NearestNeighbors._save_attributes
+    _onedal_knn_fit = NearestNeighbors._onedal_fit
+    _onedal_kneighbors = NearestNeighbors._onedal_kneighbors
+
+    def _onedal_fit(self, X, y, queue=None):
+        if sklearn_check_version("1.2"):
+            self._validate_params()
+
+        self._onedal_knn_fit(X, y, queue)
+
+        if self.contamination != "auto":
+            if not (0.0 < self.contamination <= 0.5):
+                raise ValueError(
+                    "contamination must be in (0, 0.5], " "got: %f" % self.contamination
+                )
+
+        n_samples = self.n_samples_fit_
+
+        if self.n_neighbors > n_samples:
+            warnings.warn(
+                "n_neighbors (%s) is greater than the "
+                "total number of samples (%s). n_neighbors "
+                "will be set to (n_samples - 1) for estimation."
+                % (self.n_neighbors, n_samples)
+            )
+        self.n_neighbors_ = max(1, min(self.n_neighbors, n_samples - 1))
+
+        (
+            self._distances_fit_X_,
+            _neighbors_indices_fit_X_,
+        ) = self._onedal_kneighbors(n_neighbors=self.n_neighbors_, queue=queue)
+
+        # Sklearn includes a check for float32 at this point which may not be
+        # necessary for onedal
+
+        self._lrd = self._local_reachability_density(
+            self._distances_fit_X_, _neighbors_indices_fit_X_
+        )
+
+        # Compute lof score over training samples to define offset_:
+        lrd_ratios_array = self._lrd[_neighbors_indices_fit_X_] / self._lrd[:, np.newaxis]
+
+        self.negative_outlier_factor_ = -np.mean(lrd_ratios_array, axis=1)
+
+        if self.contamination == "auto":
+            # inliers score around -1 (the higher, the less abnormal).
+            self.offset_ = -1.5
+        else:
+            self.offset_ = np.percentile(
+                self.negative_outlier_factor_, 100.0 * self.contamination
+            )
+
+        return self
+
+    def fit(self, X, y=None):
+        self._fit_validation(X, y)

```

Comment:
Move onedal_fit

Reply:
https://github.com/intel/scikit-learn-intelex/blob/main/sklearnex/neighbors/knn_unsupervised.py#L128, also exists in knn, and parameters are used in _onedal_cpu_supported and _onedal_gpu_supported :(